lib/library/ModuleLibraryPlugin.js

```diff
@@ -78,30 +125,89 @@ class ModuleLibraryPlugin extends AbstractLibraryPlugin {
 		{ options, compilation }
 	) {
 		const result = new ConcatSource(source);
-		const exportsInfo = moduleGraph.getExportsInfo(module);
+		const exportsInfos = options.export
+			? [
+					moduleGraph.getExportInfo(
+						module,
+						Array.isArray(options.export) ? options.export[0] : options.export
+					)
+				]
+			: moduleGraph.getExportsInfo(module).orderedExports;
+		const definitions =
+			/** @type {BuildMeta} */
+			(module.buildMeta).exportsFinalName || {};
+		const shortHandedExports = [];
 		const exports = [];
 		const isAsync = moduleGraph.isAsync(module);
+
 		if (isAsync) {
 			result.add(
 				`${RuntimeGlobals.exports} = await ${RuntimeGlobals.exports};\n`
 			);
 		}
-		for (const exportInfo of exportsInfo.orderedExports) {
+
+		for (const exportInfo of exportsInfos) {
 			if (!exportInfo.provided) continue;
-			const varName = `${RuntimeGlobals.exports}${Template.toIdentifier(
-				exportInfo.name
-			)}`;
-			result.add(
-				`var ${varName} = ${RuntimeGlobals.exports}${propertyAccess([
-					/** @type {string} */
-					(exportInfo.getUsedName(exportInfo.name, chunk.runtime))
-				])};\n`
+
+			let shouldContinue = false;
+
+			const reexport = exportInfo.findTarget(moduleGraph, _m => true);
+
+			if (reexport) {
+				const exp = moduleGraph.getExportsInfo(reexport.module);
+
+				for (const reexportInfo of exp.orderedExports) {
+					if (
+						reexportInfo.provided === false &&
+						reexportInfo.name === /** @type {string[]} */ (reexport.export)[0]
+					) {
+						shouldContinue = true;
+					}
+				}
+			}
+
+			if (shouldContinue) continue;
+
+			const webpackExportsProperty = exportInfo.getUsedName(
+				exportInfo.name,
+				chunk.runtime
 			);
-			exports.push(`${varName} as ${exportInfo.name}`);
+			const definition =
+				definitions[/** @type {string} */ (webpackExportsProperty)];
+			const finalName =
+				definition ||
+				`${RuntimeGlobals.exports}${Template.toIdentifier(exportInfo.name)}`;
+
+			if (!definition) {
+				result.add(
+					`var ${finalName} = ${RuntimeGlobals.exports}${propertyAccess([
+						/** @type {string} */
+						(exportInfo.getUsedName(exportInfo.name, chunk.runtime))
+					])}\n`
+				);
+			}
+
+			if (finalName && (finalName.includes(".") || finalName.includes("["))) {
+				exports.push([exportInfo.name, finalName]);
+			} else {
+				shortHandedExports.push(
+					definition && finalName === exportInfo.name
+						? finalName
+						: `${finalName} as ${exportInfo.name}`
+				);
+			}
 		}
-		if (exports.length > 0) {
-			result.add(`export { ${exports.join(", ")} };\n`);
+
+		if (shortHandedExports.length > 0) {
+			result.add(`export { ${shortHandedExports.join(", ")} };\n`);
 		}
+
+		for (const [exportName, final] of exports) {
+			result.add(
+				`export ${compilation.outputOptions.environment.const ? "const" : "var"} ${exportName} = ${final};\n`
```

Comment:
using `const` on this export will create a conflict with `info.name`  on module concatenation!
 https://github.com/webpack/webpack/blob/d98851529295d2bda276ff892af484bb10da889d/lib/optimize/ConcatenatedModule.js#L1647

This will produce multiple variables declarations with the same name. While using `var` this is not a problem, as soon as `const` is declared this will throw an error in runtime.

cc: @alexander-akait

Reply:
@bsunderhus Can you create reproducible test repo? Because we used export info and it should not happens

Reply:
> Can you create reproducible test repo?
I got this problem in my current repro (private unfortunately). I'm trying to create a public reproduction so that I can share here. I'll update here if I manage to get a reproduction

Reply:
@bsunderhus Yeah, I just need an example when it happens, no need to share full private repo

Reply:
In the meantime, here's  SS of the exports on the bundled file

![image](https://github.com/user-attachments/assets/a61a85f9-8d31-4166-be98-121102f27892)

it kinda looks like this:
```js
// I'm assuming `TelemetryHandler` here comes from `info.name`
// line 201987
// EXTERNAL MODULE: ../lib/telemetry/TelemetryHandler.js
var TelemetryHandler = __webpack_require__("hv8B");
// ...

// exportName = TelemetryHandler 
// line 202797
export const TelemetryHandler = TelemetryHandler/* TelemetryHandler */.M;
``

Reply:
@bsunderhus Can you show me a source code?

Reply:
> @bsunderhus Can you show me a source code? this module

Can't unfortunately, I'll try my best to get a public reproduction by today still 💪🏻

Reply:
Thanks, WIP on this

Reply:
Fix is there - https://github.com/webpack/webpack/pull/19459, I think tomorrow I will make release (we need to fix another issue with the module library generation)